### DISCUSSION TITLE
Getting ready for 0.16 release

### DISCUSSION THREAD

**Discussion by anitagraser** [anitagraser/movingpandas#304] [General]
This release will target users of Pandas 2.0 who need the fixes in https://github.com/movingpandas/movingpandas/commit/8ee9d7f0ab763e1226f7f6123dcf3eded7955869

I still have to 

- [x] fix the failing (minimal environment) tests and then 
- [x] merge https://github.com/movingpandas/movingpandas/pull/295 
- [x] complete #308

**Comment by anitagraser** [comment]
@raybellwaves have you had a chance to look at https://github.com/movingpandas/movingpandas/pull/295 to see if you're also happy with the solution?

**Reply by bamacgabhann** [reply]
The main thing I was unsure you'd be happy with for future compatibility was declaring the units as kwargs. Also the placement of the list of dictionaries of conversions, it could be moved out of trajectory.py. Happy to make changes if there's anything you're not sure about. 

Did you look at #303? I think that's a pandas 2.0 thing too but I'm not 100% confident it's not just my setup. If it's not just me, I can do a PR ASAP.

**Reply by bamacgabhann** [reply]
Also I can have some examples for the tutorial notebooks finished by tomorrow, I think, if you want to include those in 0.15.

**Reply by anitagraser** [reply]
> Also I can have some examples for the tutorial notebooks finished by tomorrow, I think, if you want to include those in 0.15.

That would be awesome, thanks

**Reply by anitagraser** [reply]
> the placement of the list of dictionaries of conversions, it could be moved out of trajectory.py

Right. Eventually, I think we should move that part out of trajectory.py and into a dedicated "units handling"/"conversion"/"???" utils script (similar to https://github.com/movingpandas/movingpandas/blob/main/movingpandas/geometry_utils.py)

**Reply by anitagraser** [reply]
I'm working on refactoring now

**Reply by raybellwaves** [reply]
> @raybellwaves have you had a chance to look at #295 to see if you're also happy with the solution?

I haven't. Great to see mpd have unit handling. I often have to think about whether i'm working in m/s or knots.

My only comment (and again apologies @bamacgabhann for having little time to review the code) is to check parity with naming in https://pint.readthedocs.io/en/stable/. This project is used in other projects via https://github.com/hgrecco/pint-pandas and https://github.com/xarray-contrib/pint-xarray. Either way i'm happy we have a solution. Thanks @bamacgabhann 

Now I think about this. I think pint comes from the physics community and the geo community rolled their own solution like in geopy.

**Reply by bamacgabhann** [reply]
Thanks. I am not sure if pint handles all the units mpd now can. I started by looking at geopy but quickly realised the limiting factor was the need to be able to handle whatever units the crs of a dataset might be in. Geopy doesn't handly them all. I couldn't actually find a list of CRS units anywhere, so I wrote a short script to transform a test dataset into every CRS in the proj database in turn, and extract each unique unit. That was the base unit list, and I just kept all the names which were extracted from that. I then added a few units which are used in other libraries e.g. Django, mostly derivatives of CRS units (e.g. US Survey miles). I may have cast the net too widely there - it's unlikely anyone will want to use mpd to analyse a trajectory in EPSG:4168 (unit: Gold Coast Foot) or EPSG:3167 (unit: British chain (Sears 1922 truncated)) - but I figured it would do no harm to cover even the obscure ones, and you never know. 

So that's why it doesn't necessarily match pint or geopy. I would have to have a deep think about pint and whether that's the best way to handle units in dataframes (especially as it seems to be at odds with the new pandas pyarrow backend?). I'd still class my abilities with python as very much intermediate, so it might take me a bit to get my head around it! But would be very happy to have a go at implementing it differently if pointed in the right direction.

**Reply by raybellwaves** [reply]
> Thanks. I am not sure if pint handles all the units mpd now can. I started by looking at geopy but quickly realised the limiting factor was the need to be able to handle whatever units the crs of a dataset might be in. Geopy doesn't handly them all. I couldn't actually find a list of CRS units anywhere, so I wrote a short script to transform a test dataset into every CRS in the proj database in turn, and extract each unique unit. That was the base unit list, and I just kept all the names which were extracted from that. I then added a few units which are used in other libraries e.g. Django, mostly derivatives of CRS units (e.g. US Survey miles). I may have cast the net too widely there - it's unlikely anyone will want to use mpd to analyse a trajectory in EPSG:4168 (unit: Gold Coast Foot) or EPSG:3167 (unit: British chain (Sears 1922 truncated)) - but I figured it would do no harm to cover even the obscure ones, and you never know. 
> 
> 
> 
> So that's why it doesn't necessarily match pint or geopy. I would have to have a deep think about pint and whether that's the best way to handle units in dataframes (especially as it seems to be at odds with the new pandas pyarrow backend?). I'd still class my abilities with python as very much intermediate, so it might take me a bit to get my head around it! But would be very happy to have a go at implementing it differently if pointed in the right direction. 

Thanks for sharing your thoughts. Nice work.

**Comment by anitagraser** [comment]
I'm planning to merge the #295 PR and do do the release on Monday if nothing else comes up.

**Reply by bamacgabhann** [reply]
I think I have the tutorial notebooks ready, added examples in 1 and 2. I want to look at them again tomorrow with fresher eyes for a final check, but I'll be able to push them before Monday. Should I push to #295 and keep everything together, or do a separate PR for the notebooks?

**Reply by anitagraser** [reply]
Feel free to push to #295 if you are happy with the notebooks by Monday. Otherwise, we can merge the notebook changes later.

**Reply by anitagraser** [reply]
@bamacgabhann Thank you for getting #295 ready to merge. As noted in the closing comment, there are still the topics raised in https://github.com/movingpandas/movingpandas/issues/308 regarding unit conversions in `get_length()` and `distance()` (and potentially other methods that should support this new feature. If you're already on it, let's wait with the release and include these.

**Comment by anitagraser** [comment]
I'm going through the notebooks now to see if everything there works as expected and if we should update some of the examples.

**Reply by anitagraser** [reply]
Thinking of how to handle the TimeZoneWarnings popping up with Geolife examples

**Reply by anitagraser** [reply]
```
import warnings
from movingpandas.trajectory import TimeZoneWarning
warnings.filterwarnings("ignore", category=TimeZoneWarning)
```

**Comment by anitagraser** [comment]
I've been thinking of changing the version name from 0.16rc1 to 0.16.0 to be more in line with most other libraries. Do you see any downsides to this?

**Reply by raybellwaves** [reply]
good idea

**Comment by bamacgabhann** [comment]
I have some time later this afternoon and evening if there's anything I can help with. Will look at tests and column names, if there's anything else I can do let me know.

**Reply by anitagraser** [reply]
@bamacgabhann That sounds great. I'll get 0.16.0 out now and then update the movingpandas-examples repo. We can introduce the column name automagic in 0.16.1 or 0.17

**Reply by anitagraser** [reply]
TrajectoryCollection.add_speed et al. also still need unit support https://github.com/movingpandas/movingpandas/issues/308

**Reply by anitagraser** [reply]
0.16.0 is out and I've updated the notebooks in the example repo as well as the website.
The biggest discovery was that we still need unit support in trajectory collections. Besides that, some edge cases of the improved plots don't work perfectly yet but no major issues overall 🥳
